pkg/syncer/synctarget/shard_manager.go

```diff
@@ -0,0 +1,159 @@
+	cleanupShard                     func(shardURLs workloadv1alpha1.VirtualWorkspace)
+	gvrSource                        informer.GVRSource
+}
+
+func (c *shardManager) ShardAccessForCluster(clusterName logicalcluster.Name) (ShardAccess, error) {
+	c.shardsWithStartedControllersLock.RLock()
+	defer c.shardsWithStartedControllersLock.RUnlock()
+
+	for _, shardAccess := range c.shardsWithStartedControllers {
+		if _, err := shardAccess.logicalClusterLister.ByCluster(clusterName).Get(kcpcorev1alpha1.LogicalClusterName); err == nil {
+			return shardAccess.ShardAccess, nil
+		} else if !apierrors.IsNotFound(err) {
+			return ShardAccess{}, err
+		}
+	}
+	return ShardAccess{}, apierrors.NewNotFound(logicalClusterGVR.GroupResource(), kcpcorev1alpha1.LogicalClusterName)
+}
+
+func (c *shardManager) updateShards(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) {
+	logger := klog.FromContext(ctx)
+
+	requiredShards := map[workloadv1alpha1.VirtualWorkspace]bool{}
+	if syncTarget != nil {
+		for _, shardURLs := range syncTarget.Status.VirtualWorkspaces {
+			requiredShards[shardURLs] = true
+		}
+	}
+
+	c.shardsWithStartedControllersLock.Lock()
+	defer c.shardsWithStartedControllersLock.Unlock()
+
+	// Remove obsolete controllers that don't have a shard anymore
+	for shardURLs, shard := range c.shardsWithStartedControllers {
+		if _, ok := requiredShards[shardURLs]; ok {
+			// The controllers are still expected => don't remove them
+			continue
+		}
+		// The controllers should not be running
+		// Stop them and remove it from the list of started shard controllers
+		shard.stop()
+		delete(c.shardsWithStartedControllers, shardURLs)
+	}
+
+	// Create and start missing controllers that have Virtual Workspace URLs for a shard
+	for shardURLs := range requiredShards {
+		if _, ok := c.shardsWithStartedControllers[shardURLs]; ok {
+			// The controllers are already started
+			continue
+		}
+
+		// Start the controllers
+		shardControllersContext, cancelFunc := context.WithCancel(ctx)
+		var stop context.CancelFunc = func() {
+			c.cleanupShard(shardURLs)
+			cancelFunc()
+		}
+		// Create the controllers
+		shardAccess, err := c.startShardControllers(shardControllersContext, c.gvrSource, shardURLs)
+		if err != nil {
+			logger.Error(err, "failed creating controllers for shard", "shard", shardURLs)
+			stop()
+			continue
+		}
+		informers, _ := shardAccess.SyncerDDSIF.Informers()
+		logicalClusterInformer, ok := informers[logicalClusterGVR]
+		if !ok {
+			logger.Error(nil, "informer not known and synced in the upstream syncer informer factory", "gvr", logicalClusterGVR, "shard", shardURLs)
+			stop()
+			continue
+		}
+		c.shardsWithStartedControllers[shardURLs] = shardWithStartedControllers{
+			*shardAccess, cancelFunc, logicalClusterInformer.Lister(),
+		}
+	}
+}
+
+type shardWithStartedControllers struct {
+	ShardAccess
+	stop                 context.CancelFunc
```

Comment:
surprising to see a context CancelFunc here. Either store a context (which is an anti-pattern in Go officially) or call it `func()`.

Reply:
Done